### DISCUSSION TITLE
Using typescript-resolvers mappers config with prisma models

### DISCUSSION THREAD

**Discussion by miller-productions** [dotansimha/graphql-code-generator#4101] [General] [answered]
Hi there,

Before I posted an issue, I thought I would check here via a discussion to see if there is something I'm not doing correctly.

I am attempting to use typescript-resolvers `mappers` config with generated prisma models.

As an example, I am focusing on a type I have defined named 'Task'.

From what I can see, the output from graphql-code-generator doesn't fully replace all definitions for `Task` as I would have expected.

I have written my question/confusion using the **line in bold** below.

I was hoping someone could please help me understand how mappers works and if there is something I am missing?

Here is an excerpt from my generated prisma models (`index.d.ts`)

```
/**
 * Model task
 */

export type task = {
  due_date: Date | null
  job_id: number
  priority_id: number
  status_id: number
  task_id: number
}
```

Here is the relevant section of my graphql type definitions

_I am deliberately omitting a number of fields in the definition of `Task` below, compared to what is in the model definition for `task` above, just for the purpose of this example to see how graphql-code-gen works_

```
  type Task {
    id: ID!
    priority: TaskPriority
    status: TaskStatus
  }

  type TaskPriority {
    id: ID!
    name: String 
  }

  type TaskStatus {
    id: ID!
    name: String 
  }

  type Query {
    tasks: [Task]
  }
```

Here is the relevant section of my `graphql-codegen.yml` file

```
packages/server/src/graphql/graphql-generated.ts:
    plugins:
      - "typescript"
      - "typescript-resolvers"
    config:
      useIndexSignature: true
      mappers:
        Task: ../../node_modules/.prisma/client#task
```

When I run generate, the result I get in my `graphql-generated.ts` file shows the following sections...

```
import { task } from '../../node_modules/.prisma/client';

<various code blocks omitted for brevity>

export type Query = {
   __typename?: 'Query';
  tasks?: Maybe<Array<Maybe<Task>>>;
};

export type Task = {
   __typename?: 'Task';
  id: Scalars['ID'];
  priority?: Maybe<TaskPriority>;
  status?: Maybe<TaskStatus>;
};
```

**^^^^^^ I expected this definition for `Task` here to be replaced by the model I defined in mappers.**

```
/** Mapping between all available schema types and the resolvers types */
export type ResolversTypes = ResolversObject<{
  String: ResolverTypeWrapper<Scalars['String']>,
  Boolean: ResolverTypeWrapper<Scalars['Boolean']>,
  Mutation: ResolverTypeWrapper<{}>,
  Query: ResolverTypeWrapper<{}>,
  Task: ResolverTypeWrapper<task>,
  ID: ResolverTypeWrapper<Scalars['ID']>,
  TaskPriority: ResolverTypeWrapper<TaskPriority>,
  TaskStatus: ResolverTypeWrapper<TaskStatus>,
}>;

/** Mapping between all available schema types and the resolvers parents */
export type ResolversParentTypes = ResolversObject<{
  String: Scalars['String'],
  Boolean: Scalars['Boolean'],
  Mutation: {},
  Query: {},
  Task: task,
  ID: Scalars['ID'],
  TaskPriority: TaskPriority,
  TaskStatus: TaskStatus,
}>;


Thanks for your time reading this.
Dan

**Comment by ardatan** [comment, marked as answer]
`typescript` package creates definitions according to your GraphQL Schema. Models defined in `mappers` are used in `typescript-resolvers`. 

Here you already have `task` in resolvers as I see below.
```
/** Mapping between all available schema types and the resolvers types */
export type ResolversTypes = ResolversObject<{
  String: ResolverTypeWrapper<Scalars['String']>,
  Boolean: ResolverTypeWrapper<Scalars['Boolean']>,
  Mutation: ResolverTypeWrapper<{}>,
  Query: ResolverTypeWrapper<{}>,
  Task: ResolverTypeWrapper<task>,
  ID: ResolverTypeWrapper<Scalars['ID']>,
  TaskPriority: ResolverTypeWrapper<TaskPriority>,
  TaskStatus: ResolverTypeWrapper<TaskStatus>,
}>;

/** Mapping between all available schema types and the resolvers parents */
export type ResolversParentTypes = ResolversObject<{
  String: Scalars['String'],
  Boolean: Scalars['Boolean'],
  Mutation: {},
  Query: {},
  Task: task,
  ID: Scalars['ID'],
  TaskPriority: TaskPriority,
  TaskStatus: TaskStatus,
}>;
```

**Reply by miller-productions** [reply]
Thanks @ardatan, that helps me.

I didn't pick that up when I first read through the docs but it is so obvious now that you've said it.

Much appreciated.

**Reply by dotansimha** [reply]
This article covers that: https://the-guild.dev/blog/better-type-safety-for-resolvers-with-graphql-codegen :)

**Reply by miller-productions** [reply]
Thanks again Dotan.

I've gotta say, graphql-code-generator is an awesome tool to have in the kit. 

I've only just started using it and already it is saving me so much time and pain. At the same time, it's actually helping me improve my knowledge of many other aspects of working with Typescript and GraphQL. You have really produced something excellent here. 😃
